File: android/src/main/kotlin/dev/jerrykhw/flutter_kakao_maps_sdk/FlutterKakaoMapsSDKPlugin.kt
package dev.jerrykhw.flutter_kakao_maps_sdk

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import dev.jerrykhw.flutter_kakao_maps_sdk.util.LogStreamHandler
import dev.jerrykhw.flutter_kakao_maps_sdk.view.KakaoMapViewFactory
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import java.io.InputStream

class FlutterKakaoMapsSDKPlugin : FlutterPlugin, ActivityAware {

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        pluginBinding = binding
        flutterAssets = binding.flutterAssets

        val logEventChannel = EventChannel(binding.binaryMessenger, LOG_EVENT_CHANNEL_NAME)
        logEventChannel.setStreamHandler(logStreamHandler)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) = Unit

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        val kakaoMapViewFactory =
            KakaoMapViewFactory(binding.activity, pluginBinding.binaryMessenger)

        pluginBinding.platformViewRegistry.registerViewFactory(
            KAKAO_MAP_VIEW_VIEW_ID,
            kakaoMapViewFactory
        )
    }

    override fun onDetachedFromActivityForConfigChanges() = Unit

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) = Unit

    override fun onDetachedFromActivity() = Unit

    companion object {
        private const val BASE_ID = "dev.jerrykhw.flutter_kakao_maps_sdk"

        private const val LOG_EVENT_CHANNEL_NAME = "${BASE_ID}/log"

        private const val KAKAO_MAP_VIEW_VIEW_ID = "${BASE_ID}/kakao_map_view"

        val logStreamHandler = LogStreamHandler()

        internal fun createViewMethodChannelName(id: Int): String =
            "${KAKAO_MAP_VIEW_VIEW_ID}#$id"

        private lateinit var pluginBinding: FlutterPlugin.FlutterPluginBinding
        private lateinit var flutterAssets: FlutterPlugin.FlutterAssets

        internal fun getAsset(named: String): InputStream {
            val path = flutterAssets.getAssetFilePathByName(named)
            return pluginBinding.applicationContext.assets.open(path)
        }
    }
}

File: ios/Classes/FlutterKakaoMapsSDKPlugin.swift
import Flutter

public class FlutterKakaoMapsSDKPlugin: NSObject, FlutterPlugin {
    private static var registrar: FlutterPluginRegistrar!
    
    private static let BASE_ID = "dev.jerrykhw.flutter_kakao_maps_sdk"
    
    private static let LOG_EVENT_CHANNEL_NAME = "\(BASE_ID)/log"
    
    private static let KAKAO_MAP_VIEW_VIEW_ID = "\(BASE_ID)/kakao_map_view"
    
    internal static func createViewMethodChannelName(id: Int64) -> String {
        "\(KAKAO_MAP_VIEW_VIEW_ID)#\(id)"
    }
    
    internal static let logStreamHandler = LogStreamHandler()
    
    internal static func getAssetPath(named: String) -> String {
        let key = registrar.lookupKey(forAsset: named)
        let mainBundle = Bundle.main
        let path = mainBundle.path(forResource: key, ofType: nil)!
        
        return path
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        
        let logEventChannel = FlutterEventChannel(name: LOG_EVENT_CHANNEL_NAME, binaryMessenger: registrar.messenger())
        logEventChannel.setStreamHandler(logStreamHandler)
        
        let kakaoMapViewFactory = KakaoMapViewFactory(messenger: registrar.messenger())
        registrar.register(kakaoMapViewFactory, withId: KAKAO_MAP_VIEW_VIEW_ID)
    }
}
